apps/s_client.c

```diff
@@ -3553,4 +3556,72 @@ static char *base64encode (const void *buf, size_t len)
     return out;
 }
 
+/*
+ * Host dNS Name verifier: used for checking that the hostname is in dNS format 
+ * before setting it as SNI
+ */
+static int is_dNS_name(const char *host)
+{
+    const int MAX_LABEL_LENGTH = 63;
+
+    int i;
+    int isdnsname = 0;
+    int length = strlen(host);
+    int label_length = 0;
+    int all_numeric = 1;
+
+    /*
+     * XXX: Deviation from strict DNS name syntax, also check names with '_'
+     * Check DNS name syntax, any '-' or '.' must be internal,
+     * and on either side of each '.' we can't have a '-' or '.'.
+     *
+     * If the name has just one label, we don't consider it a DNS name.  This
+     * means that "CN=sometld" cannot be precluded by DNS name constraints, but
+     * that is not a problem.
+     */
+    for (i = 0; i < length && label_length < MAX_LABEL_LENGTH; ++i) {
+        char c = host[i];
+
+        if ((c >= 'a' && c <= 'z')
+            || (c >= 'A' && c <= 'Z')
+            || c == '_') {
+            label_length += 1;
+            all_numeric = 0;
+            continue;
+        }
+
+        if (c >= '0' && c <= '9') {
+            label_length += 1;
+            continue;
+        }
+
+        /* Dot and hyphen cannot be first or last. */
+        if (i > 0 && i < length - 1) {
+            if (c == '-') {
+                label_length += 1;
+                continue;
+            }
+            /*
+             * Next to a dot the preceding and following characters must not be
+             * another dot or a hyphen.  Otherwise, record that the name is
+             * plausible, since it has two or more labels.
+             */
+            if (c == '.'
+                && host[i + 1] != '.'
+                && host[i - 1] != '-'
+                && host[i + 1] != '-') {
+                label_length = 0;
+                isdnsname = 1;
+                continue;
+            }
+        }
+        isdnsname = 0;
+        break;
+    }
```

Comment:
This seems to be duplicate code of cn2dnsid(). I'm wondering if we can't fix this without duplicating it, but I currently don't have a suggestion on how to do that.

Reply:
As per @vdukhovni 's comment on #8083 I'd be inclined to try and change this implementation to check that the provided hostname:
- contains at least two labels
- is not all-numeral

This will make it sufficiently different from cn2dnsid to justify having 2 separate methods

Reply:
I would hope that the subjectAltName dNSName and nameConstraints dNSName parsing code would already have a strict parser for DNS names. Otherwise how could they be correct?

Reply:
I'll look into it but I couldn't find anything that sounded like such a check anywhere

Reply:
NAME_CONSTRAINTS_check_CN uses cn2dnsid at its core for the hostname verification. There are also some utils for validating wildcards, but that's about as much as I could find. 

Trying to look into tests, without much luck though. 25-test_verify.t seems to be the one covering this kind of checks

Reply:
> I'd be inclined to try and change this implementation to check that the provided hostname:
>
> * contains at least two labels
>* is not all-numeral

Sounds reasonable to me.

Reply:
Note that cn2dnsid has to deal with decoding multibyte encoded CNs from the X.509 subject name, NUL characters in the name, ...  None of this applies to the hostname from the command-line.  cn2dnsid does check for at least two labels.  As for all numeric, the simpler check is to pass the value to getaddrinfo() with `AI_NUMERICHOST` set in `ai_flags`, and if it resolves, then it is an address and not a hostname.  Otherwise, it is a hostname.

We're then left with the question of which names that locally resolve to the desired IP address are appropriate for automatic use as an SNI value.  And this is not an easy question for s_client, where it is not obvious whether the intent is to probe the default certificate chain of the target server, possibly named via some local alias, not known to that server, or the intent is to verify the server's certificate against the provided name, in which case use of the name with SNI is more likely to be appropriate.

So, in fact, I'm not sure that this PR is well motivated.  Providing the server with an unexpected SNI can be worse than providing it with no SNI (often it has a certificate with many names, and SNI is not needed).  This may be especially so if s_client is not asked to abort the handshake on verification failure, or it is otherwise clear that the intent is to perform certificate verification, and not some other diagnostic task.

At the very least there would need to be a new option to prevent the automatic use of SNI, or perhaps `-servername ""` to achieve the same effect.

Reply:
It is already the default behaviour to send SNI automatically. So the motivation for this PR is to *avoid* sending "unexpected SNI". There is already an option to suppress that behaviour `-noservername`.

Reply:
I see. In that case, sure, make it tighter.